File: frontend/src/features/sessions/schemas.ts
import { z } from "zod";

import { RequestLogSchema } from "@/features/dashboard/schemas";

export const SessionModelSummarySchema = z.object({
  model: z.string(),
  requests: z.number().int().nonnegative(),
});

export const SessionAggregateSchema = z.object({
  sessionId: z.string(),
  provider: z.string(),
  useragentGroup: z.string().nullable(),
  models: z.array(SessionModelSummarySchema),
  requests: z.number().int().nonnegative(),
  inputTokens: z.number().nonnegative(),
  outputTokens: z.number().nonnegative(),
  cachedInputTokens: z.number().nonnegative(),
  costUsd: z.number(),
  errors: z.number().int().nonnegative(),
  firstSeen: z.iso.datetime({ offset: true }),
  lastSeen: z.iso.datetime({ offset: true }),
  sparkline: z.array(z.number().nonnegative()).optional(),
});

export const SessionModelBreakdownSchema = z.object({
  model: z.string(),
  requests: z.number().int().nonnegative(),
  inputTokens: z.number().nonnegative(),
  outputTokens: z.number().nonnegative(),
  cachedInputTokens: z.number().nonnegative(),
  costUsd: z.number().nullable(),
});

export const SessionsResponseSchema = z.object({
  sessions: z.array(SessionAggregateSchema),
  total: z.number().int().nonnegative(),
});

export const SessionDetailResponseSchema = z.object({
  session: SessionAggregateSchema,
  byModel: z.array(SessionModelBreakdownSchema),
  recentRequests: z.array(RequestLogSchema),
});

export const SessionAnalyticsSeriesEntrySchema = z.object({
  model: z.string(),
  reasoningEffort: z.string().nullable(),
  requests: z.number().int().nonnegative(),
  outputTokens: z.number().nonnegative(),
  cachedInputTokens: z.number().nonnegative(),
  costUsd: z.number(),
});

export const SessionAnalyticsSeatSchema = z.object({
  model: z.string(),
  reasoningEffort: z.string().nullable(),
  requests: z.number().int().nonnegative(),
  inputTokens: z.number().nonnegative(),
  outputTokens: z.number().nonnegative(),
  cachedInputTokens: z.number().nonnegative(),
  costUsd: z.number(),
  errors: z.number().int().nonnegative(),
});

export const SessionAnalyticsHistogramEntrySchema = z.object({
  label: z.string(),
  count: z.number().int().nonnegative(),
});

export const SessionAnalyticsResponseSchema = z.object({
  session: SessionAggregateSchema,
  bucketSeconds: z.number().positive(),
  series: z.array(
    z.object({
      bucketStart: z.iso.datetime({ offset: true }),
      byModel: z.array(SessionAnalyticsSeriesEntrySchema),
    }),
  ),
  seats: z.array(SessionAnalyticsSeatSchema),
  latencyHistogram: z.array(SessionAnalyticsHistogramEntrySchema),
  tokensPerRequestHistogram: z.array(SessionAnalyticsHistogramEntrySchema),
});

export type SessionAggregate = z.infer<typeof SessionAggregateSchema>;
export type SessionModelBreakdown = z.infer<typeof SessionModelBreakdownSchema>;
export type SessionDetailResponse = z.infer<typeof SessionDetailResponseSchema>;
export type SessionAnalyticsResponse = z.infer<
  typeof SessionAnalyticsResponseSchema
>;
export type SessionAnalyticsSeat = z.infer<typeof SessionAnalyticsSeatSchema>;

File: app/modules/sessions/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.request_logs.schemas import RequestLogEntry
from app.modules.shared.schemas import DashboardModel


class SessionModelRequests(DashboardModel):
    model: str
    requests: int


class SessionAggregate(DashboardModel):
    session_id: str
    provider: str
    useragent_group: str | None = None
    models: list[SessionModelRequests] = Field(default_factory=list)
    requests: int
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int
    cost_usd: float
    errors: int
    first_seen: datetime
    last_seen: datetime


class SessionListResponse(DashboardModel):
    sessions: list[SessionAggregate] = Field(default_factory=list)
    total: int


class SessionModelBreakdown(DashboardModel):
    model: str
    requests: int
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int
    cost_usd: float


class SessionDetailResponse(DashboardModel):
    session: SessionAggregate
    by_model: list[SessionModelBreakdown] = Field(default_factory=list)
    recent_requests: list[RequestLogEntry] = Field(default_factory=list)
